torch/utils/_pytree.py

```diff
@@ -26,194 +27,168 @@
+    serialize_fn: Optional[SerializeFn] = None,
+    maybe_deserialize_fn: Optional[MaybeDeserializeFn] = None,
 ) -> None:
-    if to_str_fn is None:
-        def _raise_error(spec: "TreeSpec", child_strings: List[str]) -> str:
-            raise NotImplementedError(f"Serializing {typ} not implemented")
-        to_str_fn = _raise_error
+    if serialize_fn is None:
+        def _raise_error(spec: "TreeSpec") -> Tuple[Optional[str], Optional[str]]:
+            raise NotImplementedError(f"Serializing {typ} in pytree is not registered.")
+        serialize_fn = _raise_error
 
-    if maybe_from_str_fn is None:
-        def dummy_to_str(str_spec: str) -> Optional[Tuple[Any, Context, str]]:
+    if maybe_deserialize_fn is None:
+        def dummy(type_: Optional[str], context: Optional[str]) -> Optional[Tuple[Any, Context]]:
             return None
-        maybe_from_str_fn = dummy_to_str
+        maybe_deserialize_fn = dummy
 
-    assert to_str_fn is not None
-    assert maybe_from_str_fn is not None
-    node_def = NodeDef(typ, flatten_fn, unflatten_fn, to_str_fn, maybe_from_str_fn)
+    assert serialize_fn is not None
+    assert maybe_deserialize_fn is not None
+    node_def = NodeDef(typ, flatten_fn, unflatten_fn, serialize_fn, maybe_deserialize_fn)
     SUPPORTED_NODES[typ] = node_def
 
-def _str_to_dict(str_spec: str) -> Tuple[List[str], str]:
-    assert str_spec[1] == "("
-    assert str_spec[-1] == ")"
-    context_and_child_strings = str_spec[2:-1]
-
-    child_strings = []
-    context_strings = []
-    nested_parentheses = 0
-    start_index = 0
-    for i, char in enumerate(context_and_child_strings):
-        if char == ":":
-            if nested_parentheses == 0:
-                context_strings.append(context_and_child_strings[start_index:i])
-                start_index = i + 1
-        elif char == "(":
-            nested_parentheses += 1
-        elif char == ")":
-            nested_parentheses -= 1
-
-        if nested_parentheses == 0 and char == ",":
-            child_strings.append(context_and_child_strings[start_index:i])
-            start_index = i + 1
-
-    child_strings.append(context_and_child_strings[start_index:])
-    return context_strings, ','.join(child_strings)
-
 def _dict_flatten(d: Dict[Any, Any]) -> Tuple[List[Any], Context]:
     return list(d.values()), list(d.keys())
 
 def _dict_unflatten(values: List[Any], context: Context) -> Dict[Any, Any]:
     return dict(zip(context, values))
 
-def _dict_to_str(spec: "TreeSpec", child_strings: List[str]) -> str:
+def _dict_serialize(spec: "TreeSpec") -> Tuple[Optional[str], Optional[str]]:
     assert spec.type == dict
-    context_child_strings = []
-    for key, child_string in zip(spec.context, child_strings):
-        context_child_strings.append(f"{key}:{child_string}")
-    return f"D({','.join(context_child_strings)})"
+    assert isinstance(spec.context, list)
+    serialized_context = json.dumps(spec.context)
```

Comment:
Is there a reason we need to still register a serialize for each spec individually? I was hoping we could define a single serializer that works uniformly over all contexts, under the assumption that the context is JSON dumpable. (This also raises a BC question: if someone wants to change the format of the context, how does this affect the serialization format?)

Reply:
Is it a correct assumption that the context is JSON dumpable? I feel like the context is something user defined, so it would also be pushed onto the user to determine how to serialize the context and maintain BC.

Reply:
Well no, by default something like context wouldn't have a serialization BC/FC guarantee. But on the other hand boilerplate code like this isn't really going to help you either: if someone changes the context data structure this serialization code will "just keep working" even though you've broken BC/FC.

Reply:
Another way to do this is to have users define a structured (think JSON data types only) representation for the data type that represents the stable serialization format, and then just shunt that at JSON in the schema code. Also, note that we still need to represent the *type* of the pytree node that was serialized (e.g., JSON doesn't distinguish between list and tuple)

Reply:
Do you mean to define a JSON representation for the context? How can we enforce that it's a JSON representation? 

For representing the type, if it's a custom type, can we just do `f"{CustomType.__module__}.{CustomType.__name__}"` and then for all python builtin types we just directly say `"list"`, `"tuple"`?

Reply:
yes, a fqn for the type name is a good first choice

Reply:
json representation is just any obj that is valid to pass to json.dump

Reply:
Ok, I updated the diff with a default serializer/deserializer which serializes the context w/ json. It still provides users the option to register their own serializer/deserializer.